File: packages/rust/src/network/metagraph_client.rs
//! Generic Metagraph Client for any L1 layer type
//!
//! Works with ML0 (Metagraph L0), CL1 (Currency L1), and DL1 (Data L1) nodes.
//!
//! # Example
//!
//! ```ignore
//! use constellation_sdk::network::{MetagraphClient, LayerType};
//!
//! // Connect to a Currency L1 node
//! let cl1 = MetagraphClient::new("http://localhost:9300", LayerType::CL1)?;
//!
//! // Connect to a Data L1 node
//! let dl1 = MetagraphClient::new("http://localhost:9400", LayerType::DL1)?;
//!
//! // Connect to a Metagraph L0 node
//! let ml0 = MetagraphClient::new("http://localhost:9200", LayerType::ML0)?;
//! ```

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use super::client::HttpClient;
use super::types::{
    EstimateFeeResponse, NetworkError, NetworkResult, PendingTransaction, PostDataResponse,
    PostTransactionResponse,
};
use crate::currency_types::{CurrencyTransaction, TransactionReference};
use crate::types::Signed;

/// Supported L1 layer types
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LayerType {
    /// Metagraph L0 - state channel operations
    ML0,
    /// Currency L1 - currency transactions
    CL1,
    /// Data L1 - data/update submissions
    DL1,
}

impl LayerType {
    /// Get the string representation of the layer type
    pub fn as_str(&self) -> &'static str {
        match self {
            LayerType::ML0 => "ML0",
            LayerType::CL1 => "CL1",
            LayerType::DL1 => "DL1",
        }
    }
}

impl std::fmt::Display for LayerType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

/// Cluster information from any L1 node
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClusterInfo {
    /// Cluster node count
    #[serde(default)]
    pub size: Option<u32>,
    /// Cluster ID
    #[serde(default)]
    pub cluster_id: Option<String>,
    /// Additional fields
    #[serde(flatten)]
    pub extra: HashMap<String, serde_json::Value>,
}

/// Configuration for MetagraphClient
#[derive(Debug, Clone)]
pub struct MetagraphClientConfig {
    /// Base URL of the L1 node (e.g., "http://localhost:9200")
    pub base_url: String,
    /// Layer type for API path selection
    pub layer: LayerType,
    /// Request timeout in milliseconds (default: 30000)
    pub timeout: Option<u64>,
}

/// Generic client for interacting with any Metagraph L1 layer
///
/// This client provides a unified interface for ML0, CL1, and DL1 nodes,
/// automatically selecting the correct API paths based on layer type.
///
/// # Example
///
/// ```ignore
/// use constellation_sdk::network::{MetagraphClient, LayerType};
///
/// // Connect to a Currency L1 node
/// let cl1 = MetagraphClient::new("http://localhost:9300", LayerType::CL1)?;
/// let last_ref = cl1.get_last_reference("DAG...").await?;
///
/// // Connect to a Data L1 node
/// let dl1 = MetagraphClient::new("http://localhost:9400", LayerType::DL1)?;
/// let result = dl1.post_data(&signed_data).await?;
/// ```
pub struct MetagraphClient {
    client: HttpClient,
    layer: LayerType,
}

impl MetagraphClient {
    /// Create a new MetagraphClient
    ///
    /// # Arguments
    ///
    /// * `base_url` - Base URL of the L1 node
    /// * `layer` - Layer type (ML0, CL1, or DL1)
    ///
    /// # Errors
    ///
    /// Returns an error if the HTTP client cannot be initialized
    pub fn new(base_url: impl Into<String>, layer: LayerType) -> NetworkResult<Self> {
        let client = HttpClient::new(base_url, None)?;
        Ok(Self { client, layer })
    }

    /// Create a new MetagraphClient with full configuration
    pub fn with_config(config: MetagraphClientConfig) -> NetworkResult<Self> {
        let client = HttpClient::new(config.base_url, config.timeout)?;
        Ok(Self {
            client,
            layer: config.layer,
        })
    }

    /// Get the layer type of this client
    pub fn layer(&self) -> LayerType {
        self.layer
    }

    // ============================================
    // Common operations (all layers)
    // ============================================

    /// Check the health/availability of the node
    pub async fn check_health(&self) -> bool {
        self.client
            .get::<serde_json::Value>("/cluster/info")
            .await
            .is_ok()
    }

    /// Get cluster information
    pub async fn get_cluster_info(&self) -> NetworkResult<ClusterInfo> {
        self.client.get("/cluster/info").await
    }

    // ============================================
    // Currency operations (CL1 and ML0)
    // ============================================

    /// Get the last accepted transaction reference for an address
    ///
    /// This is needed to create a new transaction that chains from
    /// the address's most recent transaction.
    ///
    /// Available on: CL1, ML0 (if currency enabled)
    ///
    /// # Errors
    ///
    /// Returns an error if called on an unsupported layer
    pub async fn get_last_reference(&self, address: &str) -> NetworkResult<TransactionReference> {
        self.assert_layer(&[LayerType::CL1, LayerType::ML0], "get_last_reference")?;
        self.client
            .get(&format!("/transactions/last-reference/{}", address))
            .await
    }

    /// Submit a signed currency transaction
    ///
    /// Available on: CL1
    ///
    /// # Errors
    ///
    /// Returns an error if called on an unsupported layer
    pub async fn post_transaction(
        &self,
        transaction: &CurrencyTransaction,
    ) -> NetworkResult<PostTransactionResponse> {
        self.assert_layer(&[LayerType::CL1], "post_transaction")?;
        self.client.post("/transactions", transaction).await
    }

    /// Get a pending transaction by hash
    ///
    /// Available on: CL1
    ///
    /// # Errors
    ///
    /// Returns an error if called on an unsupported layer
    pub async fn get_pending_transaction(
        &self,
        hash: &str,
    ) -> NetworkResult<Option<PendingTransaction>> {
        self.assert_layer(&[LayerType::CL1], "get_pending_transaction")?;
        match self.client.get(&format!("/transactions/{}", hash)).await {
            Ok(tx) => Ok(Some(tx)),
            Err(NetworkError::HttpError {
                status_code: Some(404),
                ..
            }) => Ok(None),
            Err(e) => Err(e),
        }
    }

    // ============================================
    // Data operations (DL1)
    // ============================================

    /// Estimate the fee for submitting data
    ///
    /// Available on: DL1
    ///
    /// # Errors
    ///
    /// Returns an error if called on an unsupported layer
    pub async fn estimate_fee<T: Serialize>(
        &self,
        data: &Signed<T>,
    ) -> NetworkResult<EstimateFeeResponse> {
        self.assert_layer(&[LayerType::DL1], "estimate_fee")?;
        self.client.post("/data/estimate-fee", data).await
    }

    /// Submit signed data to the Data L1 node
    ///
    /// Available on: DL1
    ///
    /// # Errors
    ///
    /// Returns an error if called on an unsupported layer
    pub async fn post_data<T: Serialize>(
        &self,
        data: &Signed<T>,
    ) -> NetworkResult<PostDataResponse> {
        self.assert_layer(&[LayerType::DL1], "post_data")?;
        self.client.post("/data", data).await
    }

    // ============================================
    // Raw HTTP access
    // ============================================

    /// Make a raw GET request to the node
    pub async fn get<T: for<'de> Deserialize<'de>>(&self, path: &str) -> NetworkResult<T> {
        self.client.get(path).await
    }

    /// Make a raw POST request to the node
    pub async fn post<T: for<'de> Deserialize<'de>, B: Serialize>(
        &self,
        path: &str,
        body: &B,
    ) -> NetworkResult<T> {
        self.client.post(path, body).await
    }

    // ============================================
    // Helpers
    // ============================================

    fn assert_layer(&self, allowed: &[LayerType], method: &str) -> NetworkResult<()> {
        if !allowed.contains(&self.layer) {
            let allowed_str: Vec<&str> = allowed.iter().map(|l| l.as_str()).collect();
            return Err(NetworkError::ConfigError(format!(
                "{}() is not available on {} layer. Available on: {}",
                method,
                self.layer,
                allowed_str.join(", ")
            )));
        }
        Ok(())
    }
}

/// Create a MetagraphClient for a specific layer
///
/// # Arguments
///
/// * `base_url` - Node URL
/// * `layer` - Layer type
///
/// # Example
///
/// ```ignore
/// use constellation_sdk::network::{create_metagraph_client, LayerType};
///
/// let client = create_metagraph_client("http://localhost:9400", LayerType::DL1)?;
/// ```
pub fn create_metagraph_client(
    base_url: impl Into<String>,
    layer: LayerType,
) -> NetworkResult<MetagraphClient> {
    MetagraphClient::new(base_url, layer)
}

File: packages/typescript/src/network/metagraph-client.ts
/**
 * Generic Metagraph Client for any L1 layer type
 *
 * Works with ML0 (Metagraph L0), CL1 (Currency L1), and DL1 (Data L1) nodes.
 *
 * @packageDocumentation
 */

import { HttpClient } from './client';
import type {
  RequestOptions,
  PendingTransaction,
  PostTransactionResponse,
  EstimateFeeResponse,
  PostDataResponse,
} from './types';
import { NetworkError } from './types';
import type {
  TransactionReference,
  CurrencyTransaction,
  TransferParams,
  TransferResult,
} from '../currency-types';
import { createCurrencyTransaction, hashCurrencyTransaction } from '../currency-transaction';
import { getPublicKeyFromPrivate, getDagAddressFromPublicKey } from '../crypto';
import type { Signed } from '../types';

/**
 * Supported L1 layer types
 */
export type LayerType = 'ml0' | 'cl1' | 'dl1';

/**
 * Configuration for MetagraphClient
 */
export interface MetagraphClientConfig {
  /** Base URL of the L1 node (e.g., 'http://localhost:9200') */
  baseUrl: string;
  /** Layer type for API path selection */
  layer: LayerType;
  /** Request timeout in milliseconds (default: 30000) */
  timeout?: number;
}

/**
 * Cluster information from any L1 node
 */
export interface ClusterInfo {
  /** Cluster node count */
  size?: number;
  /** Cluster ID */
  clusterId?: string;
  /** Additional info (varies by layer) */
  [key: string]: unknown;
}

/**
 * Generic client for interacting with any Metagraph L1 layer
 *
 * This client provides a unified interface for ML0, CL1, and DL1 nodes,
 * automatically selecting the correct API paths based on layer type.
 *
 * @example
 * ```typescript
 * // Connect to a Currency L1 node
 * const cl1 = new MetagraphClient({
 *   baseUrl: 'http://localhost:9300',
 *   layer: 'cl1'
 * });
 *
 * // Connect to a Data L1 node
 * const dl1 = new MetagraphClient({
 *   baseUrl: 'http://localhost:9400',
 *   layer: 'dl1'
 * });
 *
 * // Connect to a Metagraph L0 node
 * const ml0 = new MetagraphClient({
 *   baseUrl: 'http://localhost:9200',
 *   layer: 'ml0'
 * });
 * ```
 */
export class MetagraphClient {
  private client: HttpClient;
  private layer: LayerType;

  /**
   * Create a new MetagraphClient
   *
   * @param config - Client configuration
   */
  constructor(config: MetagraphClientConfig) {
    if (!config.baseUrl) {
      throw new Error('baseUrl is required for MetagraphClient');
    }
    if (!config.layer) {
      throw new Error('layer is required for MetagraphClient');
    }
    this.client = new HttpClient(config.baseUrl, config.timeout);
    this.layer = config.layer;
  }

  /**
   * Get the layer type of this client
   */
  getLayer(): LayerType {
    return this.layer;
  }

  // ============================================
  // Common operations (all layers)
  // ============================================

  /**
   * Check the health/availability of the node
   *
   * @param options - Request options
   * @returns True if the node is healthy
   */
  async checkHealth(options?: RequestOptions): Promise<boolean> {
    try {
      await this.client.get('/cluster/info', options);
      return true;
    } catch (error) {
      if (error instanceof NetworkError) {
        return false;
      }
      throw error;
    }
  }

  /**
   * Get cluster information
   *
   * @param options - Request options
   * @returns Cluster information
   */
  async getClusterInfo(options?: RequestOptions): Promise<ClusterInfo> {
    return this.client.get<ClusterInfo>('/cluster/info', options);
  }

  // ============================================
  // Currency operations (CL1 and ML0)
  // ============================================

  /**
   * Get the last accepted transaction reference for an address
   *
   * This is needed to create a new transaction that chains from
   * the address's most recent transaction.
   *
   * Available on: CL1, ML0 (if currency enabled)
   *
   * @param address - DAG address to query
   * @param options - Request options
   * @returns Transaction reference with hash and ordinal
   */
  async getLastReference(address: string, options?: RequestOptions): Promise<TransactionReference> {
    this.assertLayer(['cl1', 'ml0'], 'getLastReference');
    return this.client.get<TransactionReference>(
      `/transactions/last-reference/${address}`,
      options
    );
  }

  /**
   * Submit a signed currency transaction
   *
   * Available on: CL1
   *
   * @param transaction - Signed currency transaction
   * @param options - Request options
   * @returns Response containing the transaction hash
   */
  async postTransaction(
    transaction: CurrencyTransaction,
    options?: RequestOptions
  ): Promise<PostTransactionResponse> {
    this.assertLayer(['cl1'], 'postTransaction');
    return this.client.post<PostTransactionResponse>('/transactions', transaction, options);
  }

  /**
   * Get a pending transaction by hash
   *
   * Available on: CL1
   *
   * @param hash - Transaction hash
   * @param options - Request options
   * @returns Pending transaction details or null if not found
   */
  async getPendingTransaction(
    hash: string,
    options?: RequestOptions
  ): Promise<PendingTransaction | null> {
    this.assertLayer(['cl1'], 'getPendingTransaction');
    try {
      return await this.client.get<PendingTransaction>(`/transactions/${hash}`, options);
    } catch (error) {
      if (error instanceof NetworkError && error.statusCode === 404) {
        return null;
      }
      throw error;
    }
  }

  /**
   * High-level transfer: fetch last reference, create and sign transaction, submit
   *
   * Available on: CL1
   *
   * @param params - Transfer parameters (destination, amount, optional fee)
   * @param privateKey - Private key to sign with (hex string)
   * @param options - Request options, with optional lastRef for chaining
   * @returns TransferResult with hash, transaction, and reference for chaining
   *
   * @example
   * ```typescript
   * const result = await cl1.transfer(
   *   { destination: 'DAG...', amount: 100.5 },
   *   privateKey
   * );
   * console.log('Submitted:', result.hash);
   *
   * // Chain another transfer using the reference
   * const result2 = await cl1.transfer(
   *   { destination: 'DAG...2', amount: 50 },
   *   privateKey,
   *   { lastRef: result.reference }
   * );
   * ```
   */
  async transfer(
    params: TransferParams,
    privateKey: string,
    options?: RequestOptions & { lastRef?: TransactionReference }
  ): Promise<TransferResult> {
    this.assertLayer(['cl1'], 'transfer');
    const { lastRef: providedRef, ...requestOptions } = options ?? {};

    // Fetch last reference if not provided
    const sourceAddress = getSourceAddress(privateKey);
    const lastRef = providedRef ?? (await this.getLastReference(sourceAddress, requestOptions));

    // Create and sign the transaction
    const transaction = createCurrencyTransaction(params, privateKey, lastRef);

    // Submit to L1
    const response = await this.postTransaction(transaction, requestOptions);

    // Build reference for chaining
    const txHash = hashCurrencyTransaction(transaction);
    const reference: TransactionReference = {
      hash: txHash.value,
      ordinal: lastRef.ordinal + 1,
    };

    return {
      hash: response.hash,
      transaction,
      reference,
    };
  }

  /**
   * Transfer to multiple destinations sequentially with auto-chaining
   *
   * Available on: CL1
   *
   * @param transfers - Array of transfer parameters
   * @param privateKey - Private key to sign all transactions
   * @param options - Request options
   * @returns Array of TransferResults
   *
   * @example
   * ```typescript
   * const results = await cl1.transferBatch(
   *   [
   *     { destination: 'DAG...1', amount: 10 },
   *     { destination: 'DAG...2', amount: 20 },
   *   ],
   *   privateKey
   * );
   * ```
   */
  async transferBatch(
    transfers: TransferParams[],
    privateKey: string,
    options?: RequestOptions
  ): Promise<TransferResult[]> {
    this.assertLayer(['cl1'], 'transferBatch');
    const results: TransferResult[] = [];
    let lastRef: TransactionReference | undefined;

    for (const params of transfers) {
      const result = await this.transfer(params, privateKey, {
        ...options,
        lastRef,
      });
      results.push(result);
      lastRef = result.reference;
    }

    return results;
  }

  // ============================================
  // Data operations (DL1)
  // ============================================

  /**
   * Estimate the fee for submitting data
   *
   * Available on: DL1
   *
   * @param data - Signed data object to estimate fee for
   * @param options - Request options
   * @returns Fee estimate with amount and destination address
   */
  async estimateFee<T>(data: Signed<T>, options?: RequestOptions): Promise<EstimateFeeResponse> {
    this.assertLayer(['dl1'], 'estimateFee');
    return this.client.post<EstimateFeeResponse>('/data/estimate-fee', data, options);
  }

  /**
   * Submit signed data to the Data L1 node
   *
   * Available on: DL1
   *
   * @param data - Signed data object to submit
   * @param options - Request options
   * @returns Response containing the data hash
   */
  async postData<T>(data: Signed<T>, options?: RequestOptions): Promise<PostDataResponse> {
    this.assertLayer(['dl1'], 'postData');
    return this.client.post<PostDataResponse>('/data', data, options);
  }

  // ============================================
  // Raw HTTP access
  // ============================================

  /**
   * Make a raw GET request to the node
   *
   * @param path - API path
   * @param options - Request options
   * @returns Response data
   */
  async get<T>(path: string, options?: RequestOptions): Promise<T> {
    return this.client.get<T>(path, options);
  }

  /**
   * Make a raw POST request to the node
   *
   * @param path - API path
   * @param body - Request body
   * @param options - Request options
   * @returns Response data
   */
  async post<T>(path: string, body?: unknown, options?: RequestOptions): Promise<T> {
    return this.client.post<T>(path, body, options);
  }

  // ============================================
  // Helpers
  // ============================================

  private assertLayer(allowed: LayerType[], method: string): void {
    if (!allowed.includes(this.layer)) {
      throw new Error(
        `${method}() is not available on ${this.layer.toUpperCase()} layer. ` +
          `Available on: ${allowed.map((l) => l.toUpperCase()).join(', ')}`
      );
    }
  }
}

/**
 * Create a MetagraphClient for a specific layer
 *
 * @param baseUrl - Node URL
 * @param layer - Layer type
 * @param timeout - Request timeout
 */
export function createMetagraphClient(
  baseUrl: string,
  layer: LayerType,
  timeout?: number
): MetagraphClient {
  return new MetagraphClient({ baseUrl, layer, timeout });
}

/**
 * Get source DAG address from a private key
 */
function getSourceAddress(privateKey: string): string {
  const publicKey = getPublicKeyFromPrivate(privateKey);
  return getDagAddressFromPublicKey(publicKey);
}
